src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1196,45 +1238,61 @@ private HttpRequestException GetShutdownException()
+                // and perform clean up, including disposing _concurrentStreams.
+                // So if we get ObjectDisposedException here, we must have shut down the connection.
+                // Throw a retryable request exception if this is not result of some other error.
+                // This will cause retry logic to kick in and perform another connection attempt.
+                // The user should never see this exception.  See similar handling below.
+                // Throw a retryable request exception if this is not result of some other error.
+                // This will cause retry logic to kick in and perform another connection attempt.
+                // The user should never see this exception.  See also below.
+                lock (SyncObject)
                 {
-                    // We have race condition between shutting down and initiating new requests.
-                    // When we are shutting down the connection (e.g. due to receiving GOAWAY, etc)
-                    // we will wait until the stream count goes to 0, and then we will close the connetion
-                    // and perform clean up, including disposing _concurrentStreams.
-                    // So if we get ObjectDisposedException here, we must have shut down the connection.
-                    // Throw a retryable request exception if this is not result of some other error.
-                    // This will cause retry logic to kick in and perform another connection attempt.
-                    // The user should never see this exception.  See similar handling below.
-                    // Throw a retryable request exception if this is not result of some other error.
-                    // This will cause retry logic to kick in and perform another connection attempt.
-                    // The user should never see this exception.  See also below.
                     Debug.Assert(_disposed || _lastStreamId != -1);
                     Debug.Assert(_httpStreams.Count == 0);
-
-                    lock (SyncObject)
-                    {
-                        throw GetShutdownException();
-                    }
+                    throw GetShutdownException();
                 }
+            }
+
+            ArrayBuffer headerBuffer = default;
+            try
+            {
+                // Serialize headers to a temporary buffer, and do as much work to prepare to send the headers as we can
+                // before taking the write lock.
+                headerBuffer = new ArrayBuffer(InitialConnectionBufferSize, usePool: true);
+                WriteHeaders(request, ref headerBuffer);
+                ReadOnlyMemory<byte> remaining = headerBuffer.ActiveMemory;
+                Debug.Assert(remaining.Length > 0);
+
+                // Calculate the total number of bytes we're going to use (content + headers).
+                int frameCount = ((remaining.Length - 1) / FrameHeader.MaxLength) + 1;
```

Comment:
Note that `frameCount` appears to be unused now, and `totalSize` appears to be only used in tracing, and that's probably not super useful. These should be removed.

Reply:
frameCount is used to compute totalSize, and totalSize is passed to StartWriteAsync below.